### PR TITLE
ISSUE? improved arg handling

### PR DESCRIPTION
**Pull request by pnstickne**
Basically:

- All arguments must be accounted for and correctly parsed.
  This mitigates "oops" cases when arguments are not correctly
  handled by appropriate domain.

  To supply MPI_Init args (still does not address issue of late MPI_Init), use:
    `--vt_mpi_args mpiarg1 mpiarg2..`
  To supply user args (returned up through initialize):
    `-- userarg1 userarg2 ..`
  To switch back to VT args, use `--vt_args`

- Also supports 'auto pass-through' args for historic reasons.
  That is, any arguments before a "--vt_" arg including "--vt_help"
  are automatically classified as a pass-through arguments and any
  "--vt_" or argument will implicitly switch into vt-arg mode.

    `1 2 3 --vt_no_color` is the same as
    `--vt_no_color -- 1 2 3` while
    `--vt_no_color 1 2 3` is invalid (left in vt-arg context where such are invalid).

   Most notable, short vt args like '-c' or '-q' do NOT initiate
   vt-arg mode and will subject to pass-through behavior.
   (Short args have been *removed* to avoid this edge-case.)

- The syntax can also be extended indefinitely (although auto-mode only applies once at the start), for whatever use it may have:
   `1 --vt_no_color -- 2 --vt_args --vt_trace -- 3` is the same as
   `1 2 3 --vt_no_color --vt_trace` and
   `--vt_no_color --vt_trace -- 1 2 3`

   It is recommended that parameters that take values do so with the `--param=value` form.

This addresses some PEBCAK errors locally and provides a base for some future changes with directory arg handling. There is a chance similar issues are accidentally encountered on other systems.

Pass-through arguments (and the program name) are displayed in the banner for clarity.

It is a "breaking" change in that some previously valid parameters invocations are not forbidden, ie. needing to use a `--` to switch out of vt-arg context. However changes required are an artifact of the parameters supplied during invocation (as opposed to requiring code changes).

When in vt-arg context, unknown arguments will *abort the program* with an argument error.

The `--vt_help` (replaces `--help`) parameter is also more useful (and program-aborting).

### CODE REVIEW

File: src/vt/configs/arguments/args.cc [hunk 1]
```diff
@@ -486,40 +481,92 @@ namespace vt { namespace arguments {
...
+    int result = app.exit(ex);
+    // ..CLI11 returns 0 on --help, which is still terminal. Great job.
+    return result ? result : 1;
+  }
+
+  // Get the clean prog name; don't allow path bleed in usages.
+  // std::filesystem is C++17.
+  std::string prog_name = argv[0];
+  size_t l = prog_name.find_last_of("/\\");
+  if (l < prog_name.size() - 1) {
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Overflow on `""` as prog_name .. should never happen, still technically incorrect.

File: src/vt/configs/arguments/args.h [hunk 2]
```diff
@@ -147,8 +152,22 @@ struct ArgConfig {
   static std::string vt_user_str_2;
   static std::string vt_user_str_3;
 
+  /// Name of the program launched (excluding any path!)
+  static std::string prog_name;
+
+  /// Name of the program launched, aka argv[0].
+  /// Original char* objects.
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.h]: "object" (singular).

File: src/vt/configs/arguments/args.h [hunk 3]
```diff
@@ -147,8 +152,22 @@ struct ArgConfig {
...
+  static char* argv_prog_name;
+
+  /// Arguments to pass to MPI Init.
+  /// Does not include argv[0]. Original char* objects.
+  static std::vector<char*> mpi_init_args;
+  /// Arguments are being ref-returend as the result of parse(..).
+  /// Does not include argv[0]. Orignal char* objects.
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.h]: Typo.

File: src/vt/runtime/runtime.cc [hunk 4]
```diff
@@ -267,493 +271,20 @@ bool Runtime::tryFinalize() {
...
     initializeComponents();
     initializeOptionalComponents();
     initializeErrorHandlers();
     sync();
     if (theContext->getNode() == 0) {
       printStartupBanner();
+      // Enqueue a check for later in case arguments are modified before work
+      // actually executes
+      theSched->enqueue([this]{
+        this->checkForArgumentErrors();
```
Comment (pnstickne) [inline src/vt/runtime/runtime.cc]: Probably should be done up in ArgParse.. really *shouldn't* be able to modify.. not sure why comment.

File: src/vt/runtime/runtime.cc [hunk 5]
```diff
@@ -928,8 +459,23 @@ void Runtime::setup() {
   debug_print(runtime, node, "end: setup\n");
 }
 
-void Runtime::initializeContext(int argc, char** argv, MPI_Comm* comm) {
-  theContext = std::make_unique<ctx::Context>(argc, argv, is_interop_, comm);
+void Runtime::initializeContext() {
+  std::vector<char*>& mpi_args = ArgType::mpi_init_args;
```
Comment (pnstickne) [inline src/vt/runtime/runtime.cc]: `const`.

File: src/vt/runtime/runtime.cc [hunk 6]
```diff
@@ -928,8 +459,23 @@ void Runtime::setup() {
...
+  int argc = mpi_args.size() + 1;
+  char* *const argv = new char*[argc + 1];
```
Comment (pnstickne) [inline src/vt/runtime/runtime.cc]: This could be unified with code in arg parsing..

File: src/vt/configs/arguments/args.cc [hunk 7]
```diff
@@ -141,70 +144,56 @@ namespace vt { namespace arguments {
...
   auto dir    = "Name of directory to write stack files";
-  auto mod    = "Write stack dump if (node % vt_stack_mod) == 0";
-  auto g = app.add_flag("--vt_no_warn_stack",   vt_no_warn_stack,   warn);
-  auto h = app.add_flag("--vt_no_assert_stack", vt_no_assert_stack, assert);
-  auto i = app.add_flag("--vt_no_abort_stack",  vt_no_abort_stack,  abort);
-  auto j = app.add_flag("--vt_no_stack",        vt_no_stack,        stack);
-  auto k = app.add_option("--vt_stack_file",    vt_stack_file,      file, "");
-  auto l = app.add_option("--vt_stack_dir",     vt_stack_dir,       dir,  "");
-  auto m = app.add_option("--vt_stack_mod",     vt_stack_mod,       mod,  1);
+  auto mod    = "Write stack dump if (node % ArgConfig::vt_stack_mod) == 0";
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Oops - remove.

File: src/vt/configs/arguments/args.cc [hunk 8]
```diff
@@ -213,34 +202,32 @@ namespace vt { namespace arguments {
...
+    "trace_enabled)";
   auto tfile     = "Name of trace files";
   auto tdir      = "Name of directory for trace files";
-  auto tmod      = "Output trace file if (node % vt_stack_mod) == 0";
-  auto n  = app.add_flag("--vt_trace",           vt_trace,           trace);
-  auto nm = app.add_flag("--vt_trace_mpi",       vt_trace_mpi,       trace_mpi);
-  auto o  = app.add_option("--vt_trace_file",    vt_trace_file,      tfile, "");
-  auto p  = app.add_option("--vt_trace_dir",     vt_trace_dir,       tdir,  "");
-  auto q  = app.add_option("--vt_trace_mod",     vt_trace_mod,       tmod,  1);
+  auto tmod      = "Output trace file if (node % ArgConfig::vt_stack_mod) == 0";
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Oops - remove.

**Comment by lifflander** [comment]
@nlslatt Can you advise how this will affect EMPIRE? I'm worried about specializing the handling of arguments since the order in which VT is initialized with respect to other argument parsers might interact badly.

**Comment by pnstickne** [comment]
> @nlslatt Can you advise how this will affect EMPIRE? I'm worried about specializing the handling of arguments since the order in which VT is initialized with respect to other argument parsers might interact badly.

Indeed interested - these changes are to make Life Better overall and to avoid edge-case with mis-beliefs around arguments supplied; although it may require some ordering differences as-passed-to-VT. In particular, adding `--` to switch out of VT-mode and/or explicitly switching to MPI mode if any MPI parameters should be passed downward.

The overall ordering wrt MPI and startup is the same: that is, MPI_Init call remains at intact location (and it would be nice to call MPI_Init first to let it scrub all MPI-specific args, although such alters the larger contract). The caller is still guaranteed a argv/argc "cleaned" from VT arguments before the return, subject to the noted extraction differenences.

**Comment by pnstickne** [comment]
- Rebased up taking in collection and trace changes.

- Only displays help (or argument error on one node).

- Help message reads:

```
vt (Virtual Transport)

Usage:
[APP-ARGS..] [VT-ARGS] [--vt_mpi_args MPI-ARGS..] [-- APP-ARGS..]

Arguments up until the first '--vt_*' are treated as application arguments.

After the first '--vt_*', additional arguments are treated as VT arguments
unless '--vt_mpi_args' or '--' are used to switch the argument mode.
The '--vt_args' flag can be used to switch back into VT argument mode.
Modes can be switched indefinitely.

Application pass-through arguments are supplied to the host program
for further processing and are not used by VT for any configuration.
MPI (and only MPI) arguments are given directly to MPI_Init.

It is an error if an unexpected argument is encountered in VT argument mode.
The currently recognized VT arguments are listed below; availability varies
based on build and compilation settings.

```

**Comment by pnstickne** [comment]
So that's like 100% more of a help message and useful behavior than before, along with sane argument checking and a pass-through mode to hand down arguments to whatever consume changes there are.

If/as such goes in develop, it would be nice to piece up some of the runtime initialization-split changes.

**Comment by pnstickne** [comment]
Checks are "failing for reasons".

**Comment by pnstickne** [comment]
@uhetmaniuk ^_^ Is the concept of an 'anchor' transferable to argument scopes? The goal here is to *limit*, not to *add*.

Also, CLI11 1.8 has new support for sub commands; are those similar to 'anchors'? A little bit of documentation / examples would really help the ability to compare and unify.

**Comment by pnstickne** [comment]
@uhetmaniuk I think it would be beneficial to pull up the MPI_Init "prior to" returning arguments to client implementations (such that MPI_Init implementations can silently add or deduct arguments, as was the original MPI design intent). See comments in PR. Not sure if/how the goal is considered elsewhere. Such changes are no present on this branch and is considered out of scope.

**Comment by uhetmaniuk** [comment]
@pnstickne 
The issue #444 separates the creation of Runtime object and the parsing of command-line arguments (via CLI and, also, an input file from an external software).
You can see the separation in the commit 
https://github.com/DARMA-tasking/vt/commit/39b4bac3e595737098eecb4e0f1fa661acc01d79
Let me know if this is not clear.
(PS: I am in the process of cleaning-up branch #444-split-runtime-initialization into #444-split-runtime-init-cleaner).

**Comment by lifflander** [comment]
> Checks are "failing for reasons".

I'm trying to add more checks and things are a bit unstable right now. Sorry about that. Just look at Travis for now. I'm "allowed" to override checks when I merge.

**Comment by lifflander** [comment]
> @uhetmaniuk ^_^ Is the concept of an 'anchor' transferable to argument scopes? The goal here is to _limit_, such as to prevent (current and future) argument errors, not to add new functionality.
> 
> Also, CLI11 1.8 (??) has new/improved support for sub-commands; are those similar to 'anchors'? A little bit of documentation / examples would really help the ability to compare and unify.

Anchors are there to solve another problem: providing a configuration from multiple sources with a hierarchy. For instance, EMPIRE uses a YAML file and we need to allow args to come from there. Those are lower priority than command line. It also allows application-specific defaults to be specified. I think that adding those features is orthogonal to this, which is more around making the current functionality better.

Also, I agree we should switch to the new version 1.8 of CLI. They add a lot of new features

**Comment by lifflander** [comment]
@pnstickne What is the correct issue number of this?

**Comment by pnstickne** [comment]
@lifflander I've a suspicious it may 'address' https://github.com/DARMA-tasking/vt/issues/567 - in particular it is designed to rule out some issues that might cause 'unexpected' program termination, which was the original impetus.

**Comment by pnstickne** [comment]
I think it would be good to git a split-phase init (requires different loading of VT as current loading is really 'too late' for MPI args and 'too early' for additional inputs); as orthogonal, I wonder if just that can be split off from #444 - @uhetmaniuk 

That is, split-loading that does not at phase #1 deal with any loading or config changes per-se. This would require a different initialization pattern. Phase #2 would build on that with additional loading/config changes.

**Comment by lifflander** [comment]
The wrong issue number is throwing off all my analysis for the release. I'm closing this and re-opening it with the correct issue number on the branch.

**Pull request closed without merging**